src/handlers.rs

```diff
@@ -62,6 +62,39 @@ pub fn handle_request(log: &Logger, request: &protocol::Request) -> Result<Vec<u
             debug!(log, "got group"; "group" => ?group);
             serialize_group(group)
         }
+        RequestType::INITGROUPS => {
+            // initgroups is a little strange: in the public libc API, the
+            // interface is getgrouplist(), which requires that you pass one
+            // extra GID (intended to be the user's primary GID) in, which is
+            // returned as part of the result. In the glibc NSS implementation,
+            // NSS backends can implement initgroups_dyn(), which is not
+            // expected to find the primary GID (for example,
+            // _nss_files_initgroups_dyn() only looks at /etc/group);
+            // alternatively, both glibc itself and its NSCD implementation will
+            // fall back to enumerating all groups with getgrent(). It will then
+            // tack on the provided GID before returning, if it's not already in
+            // the list.
+            //
+            // There's no public API to just get the supplementary groups, so we
+            // need to get the primary group and pass it to getgrouplist()
+            // (since we don't want to implement the NSS API ourselves).
+            //
+            // One corollary is that getting supplementary groups never fails;
+            // if you ask for a nonexistent user, they just happen not to be in
+            // any groups. So the "found" value is mostly used to indicate
+            // whether the running NSCD knows about and is configured to serve
+            // initgroups responses. For that reason, we return found = 1 even
+            // if User::from_name fails.
+            let key = CStr::from_bytes_with_nul(request.key)?;
+            let user = User::from_name(key.to_str()?)?;
+            debug!(log, "got user"; "user" => ?user);
+            let groups = if let Some(user) = user {
+                getgrouplist(key, user.gid)?
```

Comment:
the comment above says to always return success, but this line implies it bails early if something goes wrong. looks like `nix` adds a too-many-groups error condition, even though libc is never going to return an error.

https://docs.rs/nix/0.20.0/nix/unistd/fn.getgrouplist.html

Reply:
might be less confusing to pull this if/else into `serialize_initgroups` and change than fn so it takes the username and primary gid

Reply:
1. I changed this specific line to log that error condition and return an empty list, and I expanded the comment to explain why we return an empty list.
2. There are still error conditions in `serialize_initgroups` that we don't handle. It's potentially important that we don't drop the connection there. However, that's a problem for _all_ of the `serialize_` functions, i.e, I think we should consider teaching `main()` not to drop the connection but instead to provide some perfunctory reply and log what's going on, because all of those cases can cause glibc to decide that there isn't a working NSCD. I think that's for another PR, and we shouldn't special-case `serialize_initgroups` here. (I also think that several of these cases can be statically proven not to be problems, if we try harder, since they're integer conversions.)

Reply:
> However, that's a problem for all of the serialize_ functions, i.e, I think we should consider teaching main() not to drop the connection but instead to provide some perfunctory reply and log what's going on, because all of those cases can cause glibc to decide that there isn't a working NSCD. I think that's for another PR, and we shouldn't special-case serialize_initgroups here.

+1